cpp/src/copying/contiguous_split.cu

```diff
@@ -762,6 +792,150 @@ struct size_of_helper {
+      size_t const ideal_num_chunks = max(size_t{1}, static_cast<size_t>(fraction * num_sms));
+
+      // make sure chunks are padded/aligned to 64 bytes.
+      size_t const chunk_size =
+        max(size_t{split_align}, _round_up_safe(bytes / ideal_num_chunks, split_align));
+      size_t const num_chunks = _round_up_safe(bytes, chunk_size) / chunk_size;
+      return thrust::pair<size_t, size_t>{num_chunks, chunk_size};
+    });
+
+  rmm::device_uvector<offset_type> chunk_offsets(num_bufs + 1, stream);
+  auto buf_count_iter = cudf::detail::make_counting_transform_iterator(
+    0, [num_bufs, num_chunks = num_chunks_func{chunks.begin()}] __device__(size_type i) {
+      return i == num_bufs ? 0 : num_chunks(i);
+    });
+  thrust::exclusive_scan(rmm::exec_policy(stream),
+                         buf_count_iter,
+                         buf_count_iter + num_bufs + 1,
+                         chunk_offsets.begin(),
+                         0);
+
+  auto out_to_in_index = [chunk_offsets = chunk_offsets.begin(), num_bufs] __device__(size_type i) {
+    return static_cast<size_type>(
+             thrust::upper_bound(thrust::seq, chunk_offsets, chunk_offsets + num_bufs + 1, i) -
+             chunk_offsets) -
+           1;
+  };
+
+  // apply the chunking.
+  auto num_chunks =
+    cudf::detail::make_counting_transform_iterator(0, num_chunks_func{chunks.begin()});
+  size_type new_buf_count =
+    thrust::reduce(rmm::exec_policy(stream), num_chunks, num_chunks + chunks.size());
+  rmm::device_uvector<dst_buf_info> d_dst_buf_info(new_buf_count, stream);
+  auto iter = thrust::make_counting_iterator(0);
+  thrust::for_each(rmm::exec_policy(stream),
+                   iter,
+                   iter + new_buf_count,
+                   [_d_dst_buf_info,
+                    d_dst_buf_info = d_dst_buf_info.begin(),
+                    chunks         = chunks.begin(),
+                    chunk_offsets  = chunk_offsets.begin(),
+                    num_bufs,
+                    num_src_bufs,
+                    out_to_in_index] __device__(size_type i) {
+                     size_type const in_buf_index = out_to_in_index(i);
+                     size_type const chunk_index  = i - chunk_offsets[in_buf_index];
+                     auto const chunk_size        = thrust::get<1>(chunks[in_buf_index]);
+                     dst_buf_info const& in       = _d_dst_buf_info[in_buf_index];
+
+                     // adjust info
+                     dst_buf_info& out = d_dst_buf_info[i];
+                     out.element_size  = in.element_size;
+                     out.value_shift   = in.value_shift;
+                     out.bit_shift     = in.bit_shift;
+                     out.valid_count =
+                       in.valid_count;  // valid count will be set to 1 if this is a validity buffer
+                     out.src_buf_index = in.src_buf_index;
+                     out.dst_buf_index = in.dst_buf_index;
```

Comment:
This seems like we want to implement `operator=` for `dst_buf_info`.

Reply:
Not necessarily.  This is only copying some of the values directly.  The remainder of them involve modifying the inputs.